### PR TITLE
Prevent elision of empty (but not nil) messages

### LINKED ISSUE #78: Bug: semantically incorrect elision of empty messages
([Originally posted here](https://github.com/protojure/protoc-plugin/issues/16#issuecomment-741501056))

[According to the proto3 spec, the default value of a message is 'not set'](https://developers.google.com/protocol-buffers/docs/proto3#default). However, [this code](https://github.com/protojure/lib/blob/master/src/protojure/protobuf/serdes/core.clj#L104) elides empty (but set) maps, essentially interpreting `nil` and `{}` as identical. On top of being generally incorrect (the receiver can't distinguish between the two despite the two being semantically different according to the spec), this has a nasty interaction with `oneof`:

```proto
syntax = "proto3";

package test;

message A {
  int32 i = 1;
}

message E {}

message M {
  oneof oo {
    A a = 1;
    E b = 2;
  }
  int32 c = 3;
}
```

```clj
user=> (require '[protojure.protobuf :as pb] '[test :as t])
nil
user=> (test/pb->M (pb/->pb (test/new-M {:oo {:a {:i 1}} :c 1})))
#test.M-record{:oo {:a #test.A-record{:i 1}}, :c 1}
user=> (test/pb->M (pb/->pb (test/new-M {:oo {:b {}} :c 1})))
#test.M-record{:oo nil, :c 1}
```

Dropping the length check in the linked code fixes the issue. I'm happy to create a PR, but I figured I'd be a little more cautious here since it seems like this check was put there intentionally and it would be good to know if there's a reason it's done that way or if it's just an oversight. (I checked the git history but this dates back to "initial public release")

### PR DESCRIPTION
**Pull request by sharrap**
Allows empty messages to be used as sentinel values inside `oneof` and for whatever other purposes people might want them for by semantically distinguishing them from `nil`/missing messages when encoding.

Resolves #78, which also goes into more detail about the rationale for this change.

### CODE REVIEW

File: src/protojure/protobuf/serdes/core.clj [hunk 1]
```diff
@@ -101,7 +101,6 @@
   (when (some? item)
     (let [data (->pb item)
           len (count data)]
-      (when-not (zero? len)
-        (.writeTag os tag 2);; embedded messages are always type=2 (string)
-        (.writeUInt32NoTag os len)
-        (.writeRawBytes os (bytes data))))))
\ No newline at end of file
+      (.writeTag os tag 2);; embedded messages are always type=2 (string)
```
Comment (ghaskins) [inline src/protojure/protobuf/serdes/core.clj:104]: Are there circumstances where some embedded types may want the (zero?) optimization?  My high level thought is I am wondering if this should be conditional, and have certain types (e.g. oneof) disable the optimization.

Comment (sharrap) [reply src/protojure/protobuf/serdes/core.clj:104]: My first thought back when I first encountered this was to have a separate `write-embedded` for `oneof` for this reason, but according to the `proto3` spec this optimization is always semantically incorrect (and therefore means if you're talking to a different protobuf implementation—e.g. in a different language—you could roundtrip a message and have them receive something different than they sent) so I opted for the more general solution.

Comment (ghaskins) [reply src/protojure/protobuf/serdes/core.clj:104]: That seems reasonable.  Worst case is, we are putting more bytes on the wire than we need to, which is the more conservative thing to do it seems.

**Review by ghaskins — APPROVED** [review]
LGTM

**Comment by ghaskins** [comment]
Your fixes are available in v1.5.12

https://github.com/protojure/lib/releases/tag/v1.5.12
https://clojars.org/protojure/versions/1.5.12

Thank you!

**Pull request merged**